**Step 1: Data bilan tanishish**
- CSV faylni ochib, birinchi 10 qatorni ko‘rish.
- Column nomlari va sonini tekshirish.
- Har bir columndagi missing values sonini aniqlash.
- Data turlari (string, numeric, date, json) haqida tushuncha hosil qilish.

In [64]:
import pandas as pd

In [65]:
df = pd.read_csv('super_dirty_students.csv')
df.head(10)

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,attendance,status,gpa,remarks,money_spent,event_time,address_raw,profile_json
0,1,Claudia Short,20,NaN,NaN,+1-619-379-4152x102,Katieland,someonegmail.com,1662247364,Data Science,NaN,active,3.72,good,$135,1629312830,"Apartment 37, South Kevin district, Tashkent, ...","{'hobbies': ['gun', 'nice'], 'skills': {'tech'..."
1,2,NaN,20,Female,ninety,NaN,Dawnburgh,psmith@chen.com,2017/08/29,Data Science,NaN,active,1.88,excellent,$152,11/10/2001 04:19 AM,UZ 100332 Tashkent South Patricia,"{hobbies:['against', 'good']}"
2,3,Kathryn Moyer,20,NaN,ninety,NaN,Lake Stevenmouth,NaN,2017-08-14,DATA SCIENCE,NaN,pending,NaN,excellent,185.0,1657837622,"Wendyshire 12-kv, dom 1, Tashkent","{'hobbies': ['fast', 'clearly'], 'skills': {'t..."
3,4,Ruben Wilson,twenty,fmale,81,NaN,Port Pamelafort,special,1973-09-17,data-sciens,NaN,inactive,0,good,175 USD,1682795130,"Apartment 16, North Tamara district, Tashkent,...","{'hobbies': ['left', 'role'], 'skills': {'tech..."
4,5,Robert Pruitt,20,Female,NaN,001-182-659-5631x02803,Kingburgh,six,2023/11/25,data_sciense,NaN,pending,4.37,average,$282,2013-10-13,"BROKEN,ADDRESS,DATA,,,","{'hobbies': ['sort', 'science'], 'skills': {'t..."
5,6,David Martinez,20,Male,ninety,001-074-828-6016x937,Hugheshaven,nathangibbs@hotmail.com,1603493647,DATA SCIENCE,110,pending,15.5,average,"85,00",1983-04-30T01:46:36,"Apartment 42, North Jamesfurt district, Tashke...","{hobbies:['order', 'keep']}"
6,7,NaN,18 years,FEMALE,71,314.018.0928x3134,Sharpfurt,nguyenlisa@@horn.org,1982-01-16,data science,NaN,active,0,excellent,275.0,19/11/2004 11:42 PM,"Lake Tracishire 29-kv, dom 8, Tashkent","{'hobbies': ['fall', 'teacher'], 'skills': {'t..."
7,8,Nicholas Dennis,21,FEMALE,ninety,8091556011,Matthewfort,avoid,2000.06.26,python,63,active,four point five,excellent,135,1978/09/23,"70576 Butler Harbor Suite 847, Janetberg",INVALID_JSON_DATA
8,9,NaN,twenty,fmale,77,854-964-5423,South Paulchester,stacysmith@taylor.biz,2000-03-15,ds,110%,active,15.5,excellent,104 USD,1995-10-20,"Williamsborough 2-kv, dom 18, Tashkent","{'hobbies': ['require', 'open'], 'skills': {'t..."
9,10,Elizabeth Villegas,22,femlae,NaN,(205)467-4476x0583,South Dianatown,anthonywalker@@gmail.com,2013-02-11T07:12:29,data_sciense,99,inactive,"3,7",average,"215,00",2014-12-16T06:44:51,"BROKEN,ADDRESS,DATA,,,","{'hobbies': ['management', 'to'], 'skills': {'..."


In [66]:
print('The number of columns:', len(df.columns))
print('Columns:', df.columns.tolist())

The number of columns: 18
Columns: ['student_id', 'name', 'age', 'gender', 'score', 'phone', 'city', 'email', 'date_of_join', 'course', 'attendance', 'status', 'gpa', 'remarks', 'money_spent', 'event_time', 'address_raw', 'profile_json']


In [67]:
df.isna().sum()

student_id        0
name            335
age             108
gender          219
score           246
phone           371
city              0
email           151
date_of_join      0
course            0
attendance      326
status            0
gpa             141
remarks         204
money_spent       0
event_time        0
address_raw       0
profile_json      0
dtype: int64

**Step 2: Oddiy string tozalash**
- Har bir string columnning bo‘sh joylarini olib tashlash (strip).
- Bo‘sh stringlarni NaN yoki Null bilan almashtirish.

In [68]:
str_cols = df.select_dtypes(include='str').columns
df[str_cols] = df[str_cols].apply(lambda x: x.str.strip())
df.head()

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,attendance,status,gpa,remarks,money_spent,event_time,address_raw,profile_json
0,1,Claudia Short,20,NaN,NaN,+1-619-379-4152x102,Katieland,someonegmail.com,1662247364,Data Science,NaN,active,3.72,good,$135,1629312830,"Apartment 37, South Kevin district, Tashkent, ...","{'hobbies': ['gun', 'nice'], 'skills': {'tech'..."
1,2,NaN,20,Female,ninety,NaN,Dawnburgh,psmith@chen.com,2017/08/29,Data Science,NaN,active,1.88,excellent,$152,11/10/2001 04:19 AM,UZ 100332 Tashkent South Patricia,"{hobbies:['against', 'good']}"
2,3,Kathryn Moyer,20,NaN,ninety,NaN,Lake Stevenmouth,NaN,2017-08-14,DATA SCIENCE,NaN,pending,NaN,excellent,185.0,1657837622,"Wendyshire 12-kv, dom 1, Tashkent","{'hobbies': ['fast', 'clearly'], 'skills': {'t..."
3,4,Ruben Wilson,twenty,fmale,81,NaN,Port Pamelafort,special,1973-09-17,data-sciens,NaN,inactive,0,good,175 USD,1682795130,"Apartment 16, North Tamara district, Tashkent,...","{'hobbies': ['left', 'role'], 'skills': {'tech..."
4,5,Robert Pruitt,20,Female,NaN,001-182-659-5631x02803,Kingburgh,six,2023/11/25,data_sciense,NaN,pending,4.37,average,$282,2013-10-13,"BROKEN,ADDRESS,DATA,,,","{'hobbies': ['sort', 'science'], 'skills': {'t..."


In [69]:
import numpy as np
df = df.replace("", np.nan)
df

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,attendance,status,gpa,remarks,money_spent,event_time,address_raw,profile_json
0,1,Claudia Short,20,NaN,NaN,+1-619-379-4152x102,Katieland,someonegmail.com,1662247364,Data Science,NaN,active,3.72,good,$135,1629312830,"Apartment 37, South Kevin district, Tashkent, ...","{'hobbies': ['gun', 'nice'], 'skills': {'tech'..."
1,2,NaN,20,Female,ninety,NaN,Dawnburgh,psmith@chen.com,2017/08/29,Data Science,NaN,active,1.88,excellent,$152,11/10/2001 04:19 AM,UZ 100332 Tashkent South Patricia,"{hobbies:['against', 'good']}"
2,3,Kathryn Moyer,20,NaN,ninety,NaN,Lake Stevenmouth,NaN,2017-08-14,DATA SCIENCE,NaN,pending,NaN,excellent,185.0,1657837622,"Wendyshire 12-kv, dom 1, Tashkent","{'hobbies': ['fast', 'clearly'], 'skills': {'t..."
3,4,Ruben Wilson,twenty,fmale,81,NaN,Port Pamelafort,special,1973-09-17,data-sciens,NaN,inactive,0,good,175 USD,1682795130,"Apartment 16, North Tamara district, Tashkent,...","{'hobbies': ['left', 'role'], 'skills': {'tech..."
4,5,Robert Pruitt,20,Female,NaN,001-182-659-5631x02803,Kingburgh,six,2023/11/25,data_sciense,NaN,pending,4.37,average,$282,2013-10-13,"BROKEN,ADDRESS,DATA,,,","{'hobbies': ['sort', 'science'], 'skills': {'t..."
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,996,Gary Sawyer,22,fmale,NaN,NaN,Gabriellefurt,amy97@woodard.net,1992.11.13,DATA SCIENCE,NaN,inactive,NaN,excellent,$64,14/10/2007 06:53 PM,"Apartment 12, Laurenhaven district, Tashkent, ...","{'hobbies': ['collection', 'his'], 'skills': {..."
996,997,Scott Clark,20,NaN,NaN,001-205-065-8737,Rogerborough,katiepage@hensley.com,2024.10.05,DATA SCIENCE,NaN,pending,"3,7",NaN,143,2021-06-21T07:46:27,"Patrickmouth 4-kv, dom 6, Tashkent",INVALID_JSON_DATA
997,998,NaN,20,MALE,NaN,264-577-8585,South Mary,someonegmail.com,19/11/2021 02:23 AM,data_sciense,95,inactive,-2,excellent,79.0,1975/05/11,"Apartment 38, New Brendan district, Tashkent, ...","{'hobbies': ['street', 'read'], 'skills': {'te..."
998,999,Brittany Barrett,21,FEMALE,86,001-663-118-1327x207,Port Rachael,richardgreen@@shannon-jenkins.info,1625723463,data_sciense,110%,active,3.06,average,"238,00",2015.05.24,"25998 Martinez Grove Apt. 473, West Scott","{'hobbies': ['game', 'total'], 'skills': {'tec..."


**Step 3: Raqam va sanalarni tozalash**
- Age, score, GPA, attendance, money_spent kabi numeric columnlarni tozalash va int yoki float formatiga
keltirish.
- Date columnlarni pandas datetime formatiga o‘tkazish.

In [70]:
df[['age', 'score', 'gpa', 'attendance', 'money_spent']]

,age,score,gpa,attendance,money_spent
0,20,NaN,3.72,NaN,$135
1,20,ninety,1.88,NaN,$152
2,20,ninety,NaN,NaN,185.0
3,twenty,81,0,NaN,175 USD
4,20,NaN,4.37,NaN,$282
...,...,...,...,...,...
995,22,NaN,NaN,NaN,$64
996,20,NaN,"3,7",NaN,143
997,20,NaN,-2,95,79.0
998,21,86,3.06,110%,"238,00"


In [71]:
num_map = {
    'twenty': '20',
    'ninety': '90',
    'four point five': '4.5',
    '18 years': '18'
}

for c in ['age', 'score', 'gpa', 'attendance', 'money_spent']:
    s = df[c].astype("str").str.strip().str.lower().replace(num_map)
    s = (
        s.str.replace('usd', '')
        .str.replace('$', '')
        .str.replace(',', '.')
        .str.replace('%', '')
        .str.replace("-", "")
    )

    df[c] = pd.to_numeric(s, errors='coerce')

for c in ['age', 'score', 'attendance']:
    df[c] = df[c].round().astype("Int64")

for c in ['gpa', 'money_spent']:
    df[c] = df[c].astype("Float64")

df.loc[df['gpa'] > 5, 'gpa'] = np.nan
df.loc[df['attendance'] > 100, 'attendance'] = 100

print(df[['age', 'score', 'gpa', 'attendance', 'money_spent']].dtypes)
df[['age', 'score', 'gpa', 'attendance', 'money_spent']].head()

age              Int64
score            Int64
gpa            Float64
attendance       Int64
money_spent    Float64
dtype: object


,age,score,gpa,attendance,money_spent
0,20,<NA>,3.72,<NA>,135.0
1,20,90,1.88,<NA>,152.0
2,20,90,<NA>,<NA>,185.0
3,20,81,0.0,<NA>,175.0
4,20,<NA>,4.37,<NA>,282.0


In [72]:
df[["date_of_join", "event_time"]]

,date_of_join,event_time
0,1662247364,1629312830
1,2017/08/29,11/10/2001 04:19 AM
2,2017-08-14,1657837622
3,1973-09-17,1682795130
4,2023/11/25,2013-10-13
...,...,...
995,1992.11.13,14/10/2007 06:53 PM
996,2024.10.05,2021-06-21T07:46:27
997,19/11/2021 02:23 AM,1975/05/11
998,1625723463,2015.05.24


In [73]:
for c in ["date_of_join", "event_time"]:
    
    s = df[c].astype('str').str.strip()
    is_unix = s.str.match(r"^\d{10}$")
    new_col = pd.Series([pd.NaT] * len(df), index=df.index)
    
    new_col[is_unix] = pd.to_datetime(s[is_unix].astype("Int64"), unit='s', errors='coerce')

    new_col[~is_unix] = pd.to_datetime(df.loc[~is_unix, c], format='mixed', errors='coerce')

    df[c] = new_col

df[["date_of_join", "event_time"]].head()


,date_of_join,event_time
0,2022-09-03 23:22:44,2021-08-18 18:53:50
1,2017-08-29 00:00:00,2001-11-10 04:19:00
2,2017-08-14 00:00:00,2022-07-14 22:27:02
3,1973-09-17 00:00:00,2023-04-29 19:05:30
4,2023-11-25 00:00:00,2013-10-13 00:00:00


**Step 4: Email va phone validation**
- Emaillarni kichik harflarga o‘tkazish va noto‘g‘ri formatdagi emaillarni aniqlash.
- Telefon raqamlarni tozalash va yagona standart formatga keltirish (masalan: +998...).

In [74]:
df['email'] = df['email'].astype('str').str.strip().str.lower()
email_pattern = r"^[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]"
valid_email = df['email'].str.match(email_pattern, na=False)
print(df.loc[~valid_email, 'email'].head(10))
df.loc[~valid_email, 'email'] = "Incorrect email format"

0             someonegmail.com
2                          NaN
3                      special
4                          six
6         nguyenlisa@@horn.org
7                        avoid
9     anthonywalker@@gmail.com
10                       stand
11          annalee@@gmail.com
12            someonegmail.com
Name: email, dtype: str


In [75]:
import re

def clean_phone(x):
    if pd.isna(x):
        return pd.NA
    s = str(x).strip()
    if s == '':
        return pd.NA
    
    s = re.sub(r"x\s*\d+$", '', s, flags=re.IGNORECASE)
    d = re.sub(r"\D", '', s)

    if len(d) == 11 and d.startswith('1'):
        d = d[1:]
    elif len(d) == 13 and d.startswith('001'):
        d = d[3:]
    elif len(d) == 10:
        d = d
    else:
        return pd.NA
    
    return f"+1-{d[:3]}-{d[3:6]}-{d[6:]}"


df['phone'] = df['phone'].apply(clean_phone)
df[['name', 'phone']].head()

,name,phone
0,Claudia Short,+1-619-379-4152
1,NaN,NaN
2,Kathryn Moyer,NaN
3,Ruben Wilson,NaN
4,Robert Pruitt,+1-182-659-5631


**Step 5: JSON parsing**
- JSON columnlarini ochish (profile_json).
- JSON ichidagi ma’lumotlarni alohida columnlarga ajratish: hobbies, skills, family, devices.
- JSON ichidagi array yoki nested objectlarni tekshirish va flatten qilish.

In [76]:
df[['name', 'profile_json']]

,name,profile_json
0,Claudia Short,"{'hobbies': ['gun', 'nice'], 'skills': {'tech'..."
1,NaN,"{hobbies:['against', 'good']}"
2,Kathryn Moyer,"{'hobbies': ['fast', 'clearly'], 'skills': {'t..."
3,Ruben Wilson,"{'hobbies': ['left', 'role'], 'skills': {'tech..."
4,Robert Pruitt,"{'hobbies': ['sort', 'science'], 'skills': {'t..."
...,...,...
995,Gary Sawyer,"{'hobbies': ['collection', 'his'], 'skills': {..."
996,Scott Clark,INVALID_JSON_DATA
997,NaN,"{'hobbies': ['street', 'read'], 'skills': {'te..."
998,Brittany Barrett,"{'hobbies': ['game', 'total'], 'skills': {'tec..."


In [77]:
import ast
import json

def parse_json(x):
    if pd.isna(x):
        return None
    s = str(x).strip()
    if s == '' or s == 'INVALID_JSON_DATA':
        return None
    
    try:
        return json.loads(s)
    except:
        pass

    try:
        return ast.literal_eval(s)
    except:
        return None
    
df['profile_obj'] = df['profile_json'].apply(parse_json)

df['hobbies'] = df['profile_obj'].apply(lambda d: d.get('hobbies') if isinstance(d, dict) else None)
df['skills'] = df['profile_obj'].apply(lambda d: d.get('skills') if isinstance(d, dict) else None)
df['family'] = df['profile_obj'].apply(lambda d: d.get('family') if isinstance(d, dict) else None)
df['devices'] = df['profile_obj'].apply(lambda d: d.get('devices') if isinstance(d, dict) else None)

df['hobbies_flat'] = df['hobbies'].apply(lambda x: ', '.join(map(str, x)) if isinstance(x, list) else None)

skills_flat = pd.json_normalize([x if isinstance(x, dict) else {} for x in df['skills']]).add_prefix('skills_')
family_flat = pd.json_normalize([x if isinstance(x, dict) else {} for x in df['family']]).add_prefix('family_')

df = df.explode("devices", ignore_index=True)
devices_norm = pd.json_normalize(
    [x if isinstance(x, dict) else {} for x in df['devices']]
).add_prefix("devices_")
df = pd.concat([df.drop(columns=['devices']), devices_norm], axis=1)


df = pd.concat([df, skills_flat, family_flat], axis=1)


df.head()


,student_id,name,age,gender,score,phone,city,email,date_of_join,course,...,devices_type,devices_brand,devices_year,skills_soft,skills_tech.python,skills_tech.excel,skills_tech.sql,family_siblings,family_income.father,family_income.mother
0,1,Claudia Short,20,NaN,<NA>,+1-619-379-4152,Katieland,Incorrect email format,2022-09-03 23:22:44,Data Science,...,laptop,HP,2021.0,"[with, onto]",2.0,5.0,1.0,4.0,1198.0,1089.0
1,1,Claudia Short,20,NaN,<NA>,+1-619-379-4152,Katieland,Incorrect email format,2022-09-03 23:22:44,Data Science,...,phone,Xiaomi,2021.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2,NaN,20,Female,90,NaN,Dawnburgh,psmith@chen.com,2017-08-29 00:00:00,Data Science,...,NaN,NaN,NaN,"[despite, clearly]",5.0,3.0,1.0,1.0,263.0,202.0
3,3,Kathryn Moyer,20,NaN,90,NaN,Lake Stevenmouth,Incorrect email format,2017-08-14 00:00:00,DATA SCIENCE,...,laptop,HP,2020.0,"[full, series]",3.0,5.0,1.0,4.0,508.0,698.0
4,3,Kathryn Moyer,20,NaN,90,NaN,Lake Stevenmouth,Incorrect email format,2017-08-14 00:00:00,DATA SCIENCE,...,phone,Samsung,2023.0,"[whose, save]",1.0,4.0,1.0,3.0,862.0,1058.0


**Step 6: Address parsing**
- Address columnni alohida qism va key-larga ajratish: shahar, tuman, pochta kodi.
- Raw address columnni saqlab, yangi columnlar hosil qilish (addr_city, addr_district, addr_postal).

In [78]:
import re

def parse_address(address):
    if pd.isna(address):
        return {'city': pd.NA, 'district': pd.NA, 'postal_code': pd.NA}
    
    s = str(address).strip()
    if s == '' or s.upper().startswith('BROKEN'):
        return {'city': pd.NA, 'district': pd.NA, 'postal_code': pd.NA}
    
    # Postal code
    m_zip = re.search(r"\b(\d{6})\b", s)  # guruhlar qivolamiz
    postal = m_zip.group(1) if m_zip else pd.NA  # group(0) bu to'liq cell, group(1) bu birinchi group
    
    # District
    district = pd.NA
    m_dist = re.search(r"\b([a-zA-Z' -]+)\s+(district)\b", s, flags=re.IGNORECASE)  # \b \b bu so'z chegarasi boshqa so'zga ulangan bo'lsa olmaydi
    m_dist_2 = re.search(r"\bUZ\s+\d{6}\s+Tashkent\s+(.+)$", s, flags=re.IGNORECASE)
    m_dist_3 = re.search(r"^(.+)\s+\d+\s*-\s*kv\b", s)
    if m_dist:
        district = m_dist.group(1).strip()
    elif m_dist_2:
        district = m_dist_2.group(1).strip()
    elif m_dist_3:
        district = m_dist_3.group(1).strip()
    
    #shahar
    m_tash = re.search(r"\b(Tashkent)\b", s, flags=re.IGNORECASE)
    if m_tash:
        city = "Tashkent"
    else: #veguldan keyingi oxirgi qismni olamiz
        parts = [p.strip() for p in s.split(',') if p.strip()]
        if parts:
            cand = parts[-1]
            cand = re.sub(r"^\d+s*", "", cand)
            if not re.fullmatch(r"\d{6}", cand) and not cand.upper().startswith('UZ'):
                city = cand

    return {"city": city, 'district': district, 'postal_code': postal}

parsed = df['address_raw'].apply(parse_address).apply(pd.Series)
df['address_city'] = parsed['city']
df['address_district'] = parsed['district']
df['address_postal_code'] = parsed['postal_code']

df[["address_raw", "address_city", "address_district", "address_postal_code"]].head(15)


,address_raw,address_city,address_district,address_postal_code
0,"Apartment 37, South Kevin district, Tashkent, ...",Tashkent,South Kevin,100539
1,"Apartment 37, South Kevin district, Tashkent, ...",Tashkent,South Kevin,100539
2,UZ 100332 Tashkent South Patricia,Tashkent,South Patricia,100332
3,"Wendyshire 12-kv, dom 1, Tashkent",Tashkent,Wendyshire,NaN
4,"Wendyshire 12-kv, dom 1, Tashkent",Tashkent,Wendyshire,NaN
5,"Apartment 16, North Tamara district, Tashkent,...",Tashkent,North Tamara,100097
6,"Apartment 16, North Tamara district, Tashkent,...",Tashkent,North Tamara,100097
7,"BROKEN,ADDRESS,DATA,,,",NaN,NaN,NaN
8,"BROKEN,ADDRESS,DATA,,,",NaN,NaN,NaN
9,"Apartment 42, North Jamesfurt district, Tashke...",Tashkent,North Jamesfurt,100551


**Step 7: Duplicate va missing data tekshirish**
- Duplicate rowsni aniqlash va o‘chirish.
- Muhim columnlarda missing values mavjudligini aniqlash va qaror qabul qilish (fillna yoki dropna).

In [79]:
safe_cols = [c for c in df.columns if not df[c].apply(lambda x: isinstance(x, (dict, list))).any()]

duplicates = df.duplicated(subset=safe_cols, keep='first')
print(duplicates.sum())

df = df.drop_duplicates(subset=safe_cols, keep='first').reset_index(drop=True)


0


In [80]:
important_cols = ["student_id", "name", "age", "gender", "score", "email", "phone", "course", "status"]
print("\nMuhim columnlarda missinglar soni")
print(df[important_cols].isna().sum())

# student id yo'q bo'lsa olib tashlaymiz
df = df.dropna(subset=['student_id'])

# text ustunlarni Unknown bilan belgilaymiz
text_cols = ["name", "gender", "email", "phone", "course", "status"]
for c in text_cols:
    df[c] = df[c].fillna("Unknown")

print("\nTozalangandan keyin missinglar soni")
print(df[important_cols].isna().sum())


Muhim columnlarda missinglar soni
student_id      0
name          529
age           174
gender        357
score         388
email           0
phone         590
course          0
status          0
dtype: int64

Tozalangandan keyin missinglar soni
student_id      0
name            0
age           174
gender          0
score         388
email           0
phone           0
course          0
status          0
dtype: int64


**Step 8: Data normalization**
- Gender columnini standard formatga keltirish: Male / Female / Unknown.
- Course columnlarini yagona nomga keltirish: Data Science / Python / Other.
- Status columnlarini standart formatga keltirish (lower-case yoki uniform).

In [83]:
def norm_gender(x):
    g = str(x).strip().lower()

    if g.startswith('f'):
        return 'Female'
    if g.startswith('m'):
        return 'Male'
    else:
        return "Unknown"
    
df['gender'] = df['gender'].apply(norm_gender)

In [86]:
c = df["course"].astype("string").str.strip().str.lower()

ds_aliases = [
    "data science", "data_science", "data-science",
    "data sciens", "data_sciense", "ds", "d.s."
]
python_aliases = ["python", "pyhton", "pythno"]

df['course'] = "Other"
df.loc[c.isin(ds_aliases), 'course'] = "Data science"
df.loc[c.isin(python_aliases), 'course'] = "Python"

In [ ]:
# Status normalized already, Unknowns have been added.

**Step 9: Final type conversion va export**
- Barcha columnlarning to‘g‘ri type da ekanligini tekshirish: string, int, float, datetime.
- Sanalarni yagona formatga keltirish (YYYY-MM-DD HH:MM:SS).
- Tozalangan data CSV faylga saqlash (super_dirty_students_cleaned.csv).

In [89]:
df.dtypes

student_id                      int64
name                              str
age                             Int64
gender                            str
score                           Int64
phone                             str
city                              str
email                             str
date_of_join            datetime64[s]
course                            str
attendance                      Int64
status                            str
gpa                           Float64
remarks                           str
money_spent                   Float64
event_time              datetime64[s]
address_raw                       str
profile_json                      str
profile_obj                    object
hobbies                        object
skills                         object
family                         object
hobbies_flat                      str
devices_type                      str
devices_brand                     str
devices_year                  float64
skills_soft 

In [90]:
df[["date_of_join", "event_time"]]

,date_of_join,event_time
0,2022-09-03 23:22:44,2021-08-18 18:53:50
1,2022-09-03 23:22:44,2021-08-18 18:53:50
2,2017-08-29 00:00:00,2001-11-10 04:19:00
3,2017-08-14 00:00:00,2022-07-14 22:27:02
4,2017-08-14 00:00:00,2022-07-14 22:27:02
...,...,...
1589,2021-11-19 02:23:00,1975-05-11 00:00:00
1590,2021-11-19 02:23:00,1975-05-11 00:00:00
1591,2021-07-08 05:51:03,2015-05-24 00:00:00
1592,2021-07-08 05:51:03,2015-05-24 00:00:00


In [92]:
df.to_csv("cleaned_dirty_students.csv", index=False)

**Step 10: QA checks**
- Original va cleaned row sonini solishtirish.
- Missing email va phone sonini tekshirish.
- Numeric columnlar (GPA, attendance, score) qiymatlari to‘g‘ri diapazonda ekanligini tekshirish.
- Duplicate rowlar yo‘qligini tasdiqlash.

In [94]:
orig = pd.read_csv("super_dirty_students.csv")
clean = pd.read_csv("cleaned_dirty_students.csv")

print(len(orig))
print(len(clean))

1000
1594


In [100]:
missing_email = clean["email"].astype(str).str.lower().isin(['incorrect email format']).sum()
missing_phone = clean["phone"].astype(str).str.lower().isin(['unknown']).sum()
print("Missing email:", missing_email)
print("Missing phone:", missing_phone)

Missing email: 1135
Missing phone: 590


In [101]:
def range_count(df, col, low=None, high=None):
    if col not in df.columns:
        return None
    s = pd.to_numeric(df[col], errors="coerce")
    mask = pd.Series([False] * len(df))
    if low is not None:
        mask = mask | (s < low)
    if high is not None:
        mask = mask | (s > high)
    return int(mask.sum())

print("GPA out-of-range:", range_count(clean, "gpa", 0, 5))
print("Attendance out-of-range:", range_count(clean, "attendance", 0, 100))
print("Score out-of-range:", range_count(clean, "score", 0, 100))


GPA out-of-range: 0
Attendance out-of-range: 0
Score out-of-range: 0


In [103]:
safe_cols = [c for c in df.columns if not df[c].apply(lambda x: isinstance(x, (dict, list))).any()]

duplicates = df.duplicated(subset=safe_cols, keep='first')
print("Duplicates:", duplicates.sum())

Duplicates: 0
